### PR TITLE
Collect GPU index assignments from SLURM gres_detail and filter Job Analyzer GPU charts (#129)

### PR DESCRIPTION
**Pull request by lushengt-meta**
Summary:

Adds GPU index collection from the SLURM REST API's gres_detail field to the GCM pipeline, and uses it in the FAIR Job Analyzer to show only the GPUs assigned to a job (instead of all 8 GPUs on the node).

## Background:
When a job uses fewer GPUs than available on a node (e.g., --gpus-per-task=1 on an 8-GPU node), the Job Analyzer previously showed metrics for all 8 GPUs. The existing GPUS_REQUESTED field comes from TRES-PER-NODE (always 8 for the full node), not the per-task allocation. TRES_GPUS_ALLOCATED correctly reports the count (e.g., 1 ) but not which specific GPU indices are assigned.

The SLURM REST API provides gres_detail — an array of strings with exact GPU index assignments per node (e.g., "gpu:ampere:1(IDX:7)"). Verified on AVA RSC: scontrol show job <id> -d | grep GRES → GRES=gpu:ampere:1(IDX:7)

## Pipeline change (Python):
- parsing.py: Added parse_gres_gpu_indices() that parses gres_detail strings into GPU index lists. Returns a comma-separated string of indices for single-node partial-GPU jobs (e.g., "7" or "0,3,5"), None for full-node (8 GPUs) or multi-node jobs. This avoids storing unnecessary data.
- squeue.py: Added GRES_GPU_INDICES field (nullable, defaults to None) and "gres_detail" → "GRES_DETAIL" REST API mapping. Adds one string column to existing fair_job_data rows — no extra entries.
- test_parsers.py: Added 12 test cases covering single GPU, multiple GPUs, range notation, full-node, multi-node, and edge cases (empty, null, N/A).

## Job Analyzer change (Hack):
- FairJob.php: Added $gresGpuIndices property
- FAIRJobAnalyzerLatestJobInfoModule.php: Queries GRES_GPU_INDICES from fair_job_data Scuba table
- FAIRJobAnalyzerPerfAnalyzerModule.php: When gresGpuIndices is available (e.g., "7"), filters all 5 GPU ODS charts (utilization, temperature, SM util, SM occupancy, memory) to gpu=(7) with per-GPU reduceTerm. When null (full-node or multi-node), shows all GPUs with the original averaged reduceTerm.

### Scope:
- Single-node partial-GPU jobs: shows only the assigned GPUs (100% accurate)
- Single-node full-GPU jobs: shows all GPUs (unchanged, no filtering needed)
- Multi-node jobs: shows all GPUs (unchanged — gres_detail has per-node values that can't be stored in fair_job_data's 1-row-per-job format)

Reviewed By: luccabb

Differential Revision: D99787988

### CODE REVIEW

File: gcm/schemas/slurm/squeue.py [hunk 1]
```diff
@@ -75,6 +76,14 @@ class JobData(DerivedCluster):
...
     RESTARTCNT: int = parsed_field(parser=int)
     SCHEDNODES: list[str] | None = parsed_field(parser=lambda s: nodelist()(s)[0])
+    GRES_GPU_INDICES: str | None = field(
+        default=None,
+        metadata={
+            "parser": parse_gres_gpu_indices,
+            "field_name": "GRES_DETAIL",
+            "slurm_field": True,
+        },
+    )
```
Comment (luccabb) [inline gcm/schemas/slurm/squeue.py:86]: can we move to `parsed_field` here?

**Review by luccabb — CHANGES_REQUESTED** [review]

File: gcm/monitoring/slurm/parsing.py [hunk 2]
```diff
@@ -278,3 +278,48 @@ def parse_scontrol_maxnodes(v: str) -> int:
...
+    indices: list[int] = []
+    for part in idx_matches[0].split(","):
+        if "-" in part:
+            start_s, end_s = part.split("-", 1)
+            indices.extend(range(int(start_s), int(end_s) + 1))
+        else:
+            indices.append(int(part))
+
+    if len(indices) >= 8 or not indices:
+        return None
```
Comment (luccabb) [inline gcm/monitoring/slurm/parsing.py]: can we adapt this to work for nodes with > 8 GPUS?

**Review by luccabb — COMMENTED** [review]
since squeue CLI has no `GRES_DETAIL` this field would always be null, we'd need to add support for Slurm REST (`SlurmRestClient`) in the collection pipeline

**Comment by luccabb** [comment]
@claude did we miss anything?

File: gcm/schemas/slurm/squeue.py (hunk 1 above)
Comment (lushengt-meta) [reply gcm/schemas/slurm/squeue.py:86]: parsed_field doesn't support a default parameter — it delegates to field(metadata=...) without passing default. Since GRES_GPU_INDICES must default to None (it's only populated via the REST API path, not squeue CLI), I used field(default=None, metadata=...) directly. This preserves all the same metadata (parser, field_name, slurm_field) that parsed_field would set.

File: gcm/monitoring/slurm/parsing.py (hunk 2 above)
Comment (lushengt-meta) [reply gcm/monitoring/slurm/parsing.py]: Done. Removed the hardcoded len(indices) >= 8 threshold from the parser — it now always returns the parsed indices regardless of count. The filtering decision is moved to the PHP caller (FAIRJobAnalyzerPerfAnalyzerModule.php), which compares the number of GPU indices against total GPUs on the node. Added test cases for 10-GPU and 16-GPU nodes to verify.

**Comment by lushengt-meta** [comment]
> since squeue CLI has no `GRES_DETAIL` this field would always be null, we'd need to add support for Slurm REST (`SlurmRestClient`) in the collection pipeline

Correct — GRES_DETAIL is a REST API-only field and is already mapped via "gres_detail": "GRES_DETAIL" in REST_TO_SQUEUE_FIELD_MAP. The _map_job_fields method in SlurmRestClient populates it from the REST API response. For CLI-based collection, slurm_field is set to False so GRES_DETAIL is excluded from JOB_DATA_SLURM_FIELDS and won't be injected into the squeue format spec. The field defaults to None when not populated.

**Comment by lushengt-meta** [comment]
> @claude did we miss anything?

addressed issues that report by claude.

File: gcm/schemas/slurm/squeue.py (hunk 1 above)
Comment (luccabb) [reply gcm/schemas/slurm/squeue.py:86]: maybe add default support to `parsed_field`?

```
from dataclasses import MISSING

...
    default=MISSING,
...
    return cast(T, field(default=default, metadata=metadata))
```

**Comment by lushengt-meta** [comment]
Hi @luccabb, could you help review this?

**Review by luccabb — APPROVED** [review]
